Mathlib/Analysis/Convex/EGauge.lean

```diff
@@ -171,24 +191,110 @@ lemma egauge_smul_right (h : c = 0 → s.Nonempty) (x : E) :
+/-- The extended gauge of a point `(a, b)` with respect to the product of balanced sets `U` and `V`
+is equal to the maximum of the extended gauges of `a` with respect to `U`
+and `b` with respect to `V`.
+-/
+theorem egauge_prod_mk {F : Type*} [AddCommGroup F] [Module 𝕜 F] {U : Set E} {V : Set F}
+    (hU : Balanced 𝕜 U) (hV : Balanced 𝕜 V) (a : E) (b : F) :
+    egauge 𝕜 (U ×ˢ V) (a, b) = max (egauge 𝕜 U a) (egauge 𝕜 V b) := by
+  refine le_antisymm (le_of_forall_lt' fun r hr ↦ ?_) (le_egauge_prod _ _ _ _)
+  simp only [max_lt_iff, egauge_lt_iff, smul_set_prod, mk_mem_prod] at hr ⊢
+  rcases hr with ⟨⟨x, hx, hxr⟩, ⟨y, hy, hyr⟩⟩
+  cases le_total ‖x‖ ‖y‖ with
+  | inl hle => exact ⟨y, ⟨hU.smul_mono hle hx, hy⟩, hyr⟩
+  | inr hle => exact ⟨x, ⟨hx, hV.smul_mono hle hy⟩, hxr⟩
 
 theorem egauge_add_add_le {U V : Set E} (hU : Balanced 𝕜 U) (hV : Balanced 𝕜 V) (a b : E) :
     egauge 𝕜 (U + V) (a + b) ≤ max (egauge 𝕜 U a) (egauge 𝕜 V b) := by
-  refine le_of_forall_lt' fun c hc ↦ ?_
-  simp only [max_lt_iff, egauge_lt_iff] at hc ⊢
-  rcases hc with ⟨⟨x, hx, hxc⟩, ⟨y, hy, hyc⟩⟩
-  wlog hxy : ‖x‖ ≤ ‖y‖ generalizing a b x y U V
-  · simpa only [add_comm] using this hV hU b a y hy hyc x hx hxc (le_of_not_le hxy)
-  refine ⟨y, ?_, hyc⟩
-  rw [smul_add]
-  exact add_mem_add (hU.smul_mono hxy hx) hy
-
-end VectorSpace
+  rw [← egauge_prod_mk hU hV a b, ← add_image_prod]
+  exact MapsTo.egauge_le 𝕜 (LinearMap.fst 𝕜 E E + LinearMap.snd 𝕜 E E) (mapsTo_image _ _) (a, b)
+
+end NormedDivisionRing
+
+section Pi
+
+variable {𝕜 : Type*} [NormedDivisionRing 𝕜]
+  {ι : Type*} {E : ι → Type*} [∀ i, AddCommGroup (E i)] [∀ i, Module 𝕜 (E i)]
+
+/-- The extended gauge of a point `x` in an indexed product
+with respect to a product of finitely many balanced sets `U i`, `i ∈ I`,
+(and the whole spaces for the other indices)
+is the supremum of the extended gauges of the components of `x`
+with respect to the corresponding balanced set.
+
+This version assumes the following technical condition:
+- either `I` is the universal set;
+- or one of `x i`, `i ∈ I`, is nonzero;
+- or `𝕜` is nontrivially normed.
+-/
+theorem egauge_pi' {I : Set ι} (hI : I.Finite)
+    {U : ∀ i, Set (E i)} (hU : ∀ i ∈ I, Balanced 𝕜 (U i))
+    (x : ∀ i, E i) (hI₀ : I = univ ∨ (∃ i ∈ I, x i ≠ 0) ∨ (𝓝[≠] (0 : 𝕜)).NeBot) :
+    egauge 𝕜 (I.pi U) x = ⨆ i ∈ I, egauge 𝕜 (U i) (x i) := by
+  refine le_antisymm ?_ (iSup₂_le fun i hi ↦ le_egauge_pi hi _ _)
+  refine le_of_forall_lt' fun r hr ↦ ?_
+  have : ∀ i ∈ I, ∃ c : 𝕜, x i ∈ c • U i ∧ ‖c‖ₑ < r := fun i hi ↦
+    egauge_lt_iff.mp <| (le_iSup₂ i hi).trans_lt hr
+  choose! c hc hcr using this
+  obtain ⟨c₀, hc₀, hc₀I, hc₀r⟩ :
+      ∃ c₀ : 𝕜, (c₀ ≠ 0 ∨ I = univ) ∧ (∀ i ∈ I, ‖c i‖ ≤ ‖c₀‖) ∧ ‖c₀‖ₑ < r := by
+    have hr₀ : 0 < r := hr.bot_lt
+    have Hpos [hbot : (𝓝[≠] (0 : 𝕜)).NeBot] : ∃ c₀ ≠ (0 : 𝕜), ‖c₀‖ₑ < r :=
+      frequently_iff_neBot.mpr hbot |>.and_eventually
+        ((continuous_enorm.tendsto' 0 0 (by simp)).eventually_lt_const hr₀) |>.exists
```

Comment:
I think this is worth extracting to an `exists_enorm_lt` lemma with a docstring referencing `NormedField.exists_norm_lt`

Reply:
Or better yet, referencing the lemma in #23740

Reply:
I'm going to merge my PR, then you can reference the new lemma in #23740.